### PR TITLE
feat(category): add action for toggling a category filter

### PR DESCRIPTION
**Pull request by lderrickable**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the new behavior?
An action was added for toggling a filter. Simplifies a front end application by extracting management of the applied filters to daffodil

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: libs/category/src/effects/category.effects.ts [hunk 1]
```diff
@@ -135,7 +160,20 @@ export class DaffCategoryEffects {
...
+		appliedFilters: DaffCategoryFilterAction[]
+	): DaffCategoryFilterAction[] {
+		if(!appliedFilters) return [toggledFilter];
+
+		const nonMatchingFilters = appliedFilters.filter(filter => filter !== toggledFilter);
+
+		if(nonMatchingFilters.length === appliedFilters.length) nonMatchingFilters.push(toggledFilter);
+
+		return nonMatchingFilters.length ? nonMatchingFilters : null;
+	}
```
Comment (griest024) [inline libs/category/src/effects/category.effects.ts]: This method is hard to understand. At the very least it needs comments explaining what is going on and what the purpose of the method is but I would rewrite it to make the code more clear. As I understand it, this method will either add or remove the toggled filter from the applied filters based on whether or not it is already applied. Assuming my understanding is correct, a clearer implementation would be:
```ts
private toggleCategoryFilter(
  toggledFilter: DaffCategoryFilterAction, 
  appliedFilters: DaffCategoryFilterAction[] = []
): DaffCategoryFilterAction[] {
  return appliedFilters.includes(toggledFilter)
    // filter is applied, remove it
    ? appliedFilters.filter(filter => filter !== toggledFilter)
    // filter is not applied, add it
    : appliedFilters.concat([toggledFilter])
}
```

Unless there is a reason you need to return `null`, I think this is better.

**Comment by griest024** [comment]
Needs build fixed. I'm not sure changing all the state to arrays was needed; its possible that a default argument would have been sufficient. However, I do think its better to have state that is an array to be initialized to an empty array so you don't have to defensively program against it.

**Comment by lderrickable** [comment]
> Needs build fixed. I'm not sure changing all the state to arrays was needed; its possible that a default argument would have been sufficient. However, I do think its better to have state that is an array to be initialized to an empty array so you don't have to defensively program against it.

Yea, sorry. I thought it was going to be necessary, because the Magento category call blows up if you pass it anything but a valid filter. So I thought it would end up being necessary to change a lot of stuff, because that method returns an empty array instead of null. But yea, I think it ended up being unnecessary.

**Comment by griest024** [comment]
Its all good though, I think having empty arrays is better than `null`.
> the Magento category call blows up if you pass it anything but a valid filter
seems about right 🙄

**Pull request merged**